packages/web/app/src/lib/preflight-sandbox/graphiql-plugin.tsx

```diff
@@ -212,70 +126,200 @@ const UpdatePreflightScriptMutation = graphql(`
   }
 `);
 
-function stopWorker() {
-  preflightWorker.terminate(); // This kills the worker
-  preflightWorker = new PreflightWorker();
+const PreflightScript_TargetFragment = graphql(`
+  fragment PreflightScript_TargetFragment on Target {
+    id
+    preflightScript {
+      id
+      sourceCode
+    }
+  }
+`);
+
+type LogRecord = LogMessage | { type: 'separator' };
+
+export function usePreflightScript(args: {
+  target: FragmentType<typeof PreflightScript_TargetFragment> | null;
+}) {
+  const target = useFragment(PreflightScript_TargetFragment, args.target);
+  const [worker, setWorker] = useState(() => new PreflightWorker());
+  const [isPreflightScriptEnabled, setIsPreflightScriptEnabled] = useLocalStorage(
+    'hive:laboratory:isPreflightScriptEnabled',
+    false,
+  );
+  const [environmentVariables, setEnvironmentVariables] = useLocalStorage(
+    'hive:laboratory:environment',
+    '',
+  );
+  const [isRunning, setIsRunning] = useState(false);
+  const [logs, setLogs] = useState<LogRecord[]>([]);
+  const onNextFinishRef = useRef<PromiseWithResolvers<Record<string, unknown>> | undefined>();
+
+  useEffect(() => {
+    worker.onmessage = (ev: MessageEvent<PreflightScriptResult>) => {
+      if (ev.data.type === 'result') {
+        setIsRunning(false);
+        setEnvironmentVariables(JSON.stringify(ev.data.environmentVariables, null, 2));
+        setLogs(logs => [...logs, { type: 'separator' }]);
+
+        if (onNextFinishRef.current) {
+          onNextFinishRef.current.resolve(ev.data.environmentVariables);
+          onNextFinishRef.current = undefined;
+        }
+        return;
+      }
+      if (ev.data.type === 'log') {
+        const message = ev.data.message;
+        setLogs(logs => [...logs, message]);
+        return;
+      }
+      if (ev.data.type === 'error') {
+        const error = ev.data.error;
+        setLogs(logs => [...logs, error]);
+        setIsRunning(false);
+        if (onNextFinishRef.current) {
+          onNextFinishRef.current.reject(error);
+          onNextFinishRef.current = undefined;
+        }
+        return;
+      }
+
+      throw new Error('Received unexpected response from worker.');
+    };
+
+    // terminate worker when leaving laboratory
+    return () => {
+      worker.onmessage = () => {};
+      worker.terminate();
+    };
+  }, [worker]);
+
+  const abort = useCallback(
+    (isTimeout = false) => {
+      setLogs(logs => [
+        ...logs,
+        isTimeout
+          ? new Error(
+              `Preflight script execution timed out after ${PREFLIGHT_TIMEOUT / 1000} seconds`,
+            )
+          : '> Preflight script interrupted by user',
```

Comment:
on the first open edit script modal I get

https://github.com/user-attachments/assets/5e699925-e63a-446e-8895-f299b9d597f2